src/storage/aggregator.cc

```diff
@@ -0,0 +1,327 @@
+    uint32_t ch_length = 0;
+    row_view->GetString(4, &ch, &ch_length);
+    switch (aggr_col_type_) {
+        case DataType::kSmallInt:
+        case DataType::kInt:
+        case DataType::kBigInt: {
+            int64_t origin_val = *reinterpret_cast<int64_t*>(ch);
+            buffer->aggr_val_.vlong = origin_val;
+            break;
+        }
+        case DataType::kFloat: {
+            float origin_val = *reinterpret_cast<float*>(ch);
+            buffer->aggr_val_.vfloat = origin_val;
+            break;
+        }
+        case DataType::kDouble: {
+            double origin_val = *reinterpret_cast<double*>(ch);
+            buffer->aggr_val_.vdouble = origin_val;
+            break;
+        }
+        default: {
+            PDLOG(WARNING, "unsupported data type");
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Aggregator::FlushAggrBuffer(const std::string& key, const AggrBuffer& buffer) {
+    std::string encoded_row;
+    std::string aggr_val;
+    switch (aggr_col_type_) {
+        case DataType::kSmallInt:
+        case DataType::kInt:
+        case DataType::kBigInt: {
+            int64_t tmp_val = buffer.aggr_val_.vlong;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(int64_t));
+            break;
+        }
+        case DataType::kFloat: {
+            float tmp_val = buffer.aggr_val_.vfloat;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(float));
+            break;
+        }
+        case DataType::kDouble: {
+            double tmp_val = buffer.aggr_val_.vdouble;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(double));
+            break;
+        }
+        default: {
+            PDLOG(WARNING, "unsupported data type");
+            return false;
+        }
+    }
+
+    int str_length = key.size() + aggr_val.size();
+    uint32_t row_size = row_builder_.CalTotalLength(str_length);
+    encoded_row.resize(row_size);
+    row_builder_.SetBuffer(reinterpret_cast<int8_t*>(&(encoded_row[0])), row_size);
+    row_builder_.AppendString(key.c_str(), key.size());
+    row_builder_.AppendTimestamp(buffer.ts_begin_);
+    row_builder_.AppendTimestamp(buffer.ts_end_);
+    row_builder_.AppendInt32(buffer.aggr_cnt_);
+    row_builder_.AppendString(aggr_val.c_str(), aggr_val.size());
+    row_builder_.AppendInt64(buffer.binlog_offset_);
+
+    int64_t time = ::baidu::common::timer::get_micros() / 1000;
+    dimensions_.Mutable(0)->set_key(key);
+    bool ok = aggr_table_->Put(time, encoded_row, dimensions_);
+    if (!ok) {
+        LOG(ERROR) << "aggregator put failed";
+        return false;
+    }
+
+    // init the next window range
+    auto& latest_val = aggr_buffer_map_.at(key);
+    int64_t latest_ts = latest_val.ts_end_ + 1;
+    latest_val.clear();
+    latest_val.ts_begin_ = latest_ts;
+    if (window_type_ == WindowType::kRowsRange) {
+        latest_val.ts_end_ = latest_ts + window_size_ - 1;
+    }
```

Comment:
for efficiency and thread-safe, the AggrBuffer can be copied and passed in. The init/clear phases can be done outside and protected.